src/jit/morph.cpp

```diff
@@ -12913,9 +12913,11 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 {
                     noway_assert(varTypeIsIntOrI(tree));
 
-                    tree->gtOp.gtOp2 = op2 = gtNewOperNode(GT_NEG, tree->gtType, op2); // The type of the new GT_NEG
-                                                                                       // node should be the same
-                    // as the type of the tree, i.e. tree->gtType.
+                    // The type of the new GT_NEG node cannot just be op2->TypeGet().
+                    // Otherwise we may sign-extend incorrectly in cases where the NEG
+                    // node ends up feeding directly into a cast, for example (byte)(0 - s_1)
+                    // where s_1 is a byte.
+                    tree->gtOp.gtOp2 = op2 = gtNewOperNode(GT_NEG, genActualType(op2->TypeGet()), op2);
```

Comment:
I think the comment is slightly confusing: you use `byte` here in C# sense, where `byte` is unsigned. `TYP_BYTE` in the jit is signed. I think it would be more clear if you used the jit terminology, `TYP_UBYTE` in this case.

Reply:
NEG --> GT_NEG